File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    answer = 0
    data = [[0] * (len(friends)) for _ in range(len(friends))]
    gift_point = [0] * len(friends)
    for gift in gifts:
        gift = gift.split()
        giver = friends.index(gift[0])
        getter = friends.index(gift[1])
        data[giver][getter] += 1
        gift_point[giver] -= 1
        gift_point[getter] += 1
    
    res = [0] * len(friends)
    for i in range(len(friends)):
        for j in range(len(friends)):
            if i == j:
                continue
            if data[i][j] > data[j][i]:
                res[i] += 1
            elif data[i][j] == data[j][i] and gift_point[i] < gift_point[j]:
                    res[i] += 1
    
    answer = max(res)
    return answer

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    
    static int N;
    static Map<String, Integer> names = new HashMap<>();
    
    public int solution(String[] friends, String[] gifts) {
        N = friends.length;
        for (int i = 0; i < N; i++) {
            names.put(friends[i], i);
        }
        int[][] data = new int[N][N];
        
        for (String gift : gifts) {
            String[] split = gift.split(" ");
            int sender = names.get(split[0]);
            int receiver = names.get(split[1]);
            data[sender][receiver]++;
        }
        
        return solve(data);
    }
    
    private int solve(int[][] data) {
        int res = 0, temp = 0;
        
        for (int i = 0; i < N; i++) {
            temp = 0;
            for (int j = 0; j < N; j++) {
                if (data[i][j] > data[j][i]) temp++;
                else if (data[i][j] == data[j][i]) {
                    int ii = calculate(i, data);
                    int jj = calculate(j, data);
                    
                    if (ii > jj) temp++;
                }
            }
            res = Math.max(res, temp);
        }
        
        return res;
    }
    
    private int calculate(int x, int[][] data) {
        int res = 0;
        
        for (int i = 0; i < N; i++) {
            res += data[x][i];
            res -= data[i][x];
        }
        
        return res;
    }
}